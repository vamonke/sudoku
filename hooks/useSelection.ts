import { useState } from "react";

export function useSelection() {
  const [selectedIndex, setSelectedIndex] = useState<number | null>(null);

  const onFocus = (index: number) => {
    setSelectedIndex(index);
  };

  const onBlur = () => {
    setSelectedIndex(null);
  };

  return {
    selectedIndex,
    onFocus,
    onBlur,
  };
}
