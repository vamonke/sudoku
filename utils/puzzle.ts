import { supabaseClient } from "@/utils/supabase/client";

export const fetchNewPuzzle = async (puzzleId: string) => {
  const { data, error } = await supabaseClient.rpc("get_random_puzzle", {
    pid: puzzleId,
  });

  if (error) {
    throw error;
  }

  return data;
};
