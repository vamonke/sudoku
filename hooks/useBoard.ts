import { useCallback, useMemo, useState } from "react";
import { findConflict, parsePuzzleString } from "@/utils/puzzle";
import { Board } from "@/types";

export function useBoard(initialPuzzle: string) {
  const [board, setBoard] = useState<Board>(parsePuzzleString(initialPuzzle));

  const reset = useCallback(() => {
    setBoard((prevBoard) =>
      prevBoard.map((cell) => {
        if (!cell.editable) return cell;
        return { ...cell, value: null };
      })
    );
  }, []);

  const updateCell = (index: number, value: number | null) => {
    setBoard((prevBoard) =>
      prevBoard.map((cell, i) => {
        if (i !== index) return cell;
        return { ...cell, value };
      })
    );
  };

  const conflictSet = useMemo(() => findConflict(board), [board]);
  const isComplete =
    conflictSet.size === 0 && board.every((cell) => cell.value !== null);

  return {
    board,
    setBoard,
    reset,
    updateCell,
    isComplete,
    conflictSet,
  };
}
