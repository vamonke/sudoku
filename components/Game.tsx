"use client";

import { useGame } from "@/hooks/useGame";
import { SudokuPuzzle } from "@/types";
import Grid from "./Grid";
import Controls from "./Controls";

export default function Game(props: { initialPuzzle: SudokuPuzzle }) {
  const { initialPuzzle } = props;
  const {
    board,
    restart,
    handleNewGame,
    undo,
    isComplete,
    onChangeCell,
    toggleShowConflict,
    showConflict,
    conflictSet,
  } = useGame(initialPuzzle);

  return (
    <main className="flex flex-col h-full">
      <div className="p-4">
        <h1 className="text-xl text-center">SUDOKU</h1>
        <Controls
          restart={restart}
          newGame={handleNewGame}
          undo={undo}
          toggleShowConflict={toggleShowConflict}
          showConflict={showConflict}
        />
      </div>
      <div className="grow flex flex-col justify-center items-center pb-8">
        <div className="aspect-square" style={{ height: 512, width: 512 }}>
          {board ? (
            <Grid
              board={board}
              onChangeCell={onChangeCell}
              isComplete={isComplete}
              showConflict={showConflict}
              conflictSet={conflictSet}
            />
          ) : (
            <EmptyGrid />
          )}
        </div>
      </div>
      <div className="flex flex-row gap-4 justify-center items-center p-4">
        <div>
          Use arrow keys to navigate. Use numbers 1-9 to fill in the cells.
        </div>
      </div>
    </main>
  );
}

const EmptyGrid = () => {
  const cells = new Array(81).fill(null);
  return (
    <div className="grid grid-cols-9 grid-rows-9 gap-1 h-full w-full">
      {cells.map((_, index) => (
        <div key={index} className="bg-gray-200" />
      ))}
    </div>
  );
};
