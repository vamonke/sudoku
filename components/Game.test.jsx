import { mockBoard } from "@/utils/mocks";
import "@testing-library/jest-dom";
import { render } from "@testing-library/react";
import Game from "./Game";

jest.mock("../hooks/useGame", () => ({
  useGame: jest.fn(),
}));

import { useGame } from "../hooks/useGame";

describe("Game", () => {
  describe("renders correctly", () => {
    it("when board is null", () => {
      useGame.mockImplementation(() => ({ board: mockBoard }));
      const { container } = render(<Game />);
      expect(container).toMatchSnapshot();
    });
    it("when board is not null", () => {
      useGame.mockImplementation(() => ({ board: null }));
      const { container } = render(<Game />);
      expect(container).toMatchSnapshot();
    });
  });
});
