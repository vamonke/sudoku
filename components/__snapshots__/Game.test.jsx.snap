// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game renders correctly when board is not null 1`] = `
<div>
  <main
    class="flex flex-col h-full bg-indigo-50 p-4"
  >
    <div
      class="p-2"
    >
      <h1
        class="text-5xl text-center font-semibold text-indigo-800"
      >
        SUDOKU
      </h1>
      <div
        class="text-center text-gray-600 mt-3"
      >
        Fill the 9x9 grid with digits so that each column, row, and 3x3 section contains all the digits from 1 to 9.
      </div>
    </div>
    <div
      class="grow flex flex-col justify-center items-center mt-4 mb-8"
    >
      <div
        class="aspect-square h-full"
        style="max-height: 75vh; max-width: 80vw;"
      >
        <div
          class="grid grid-cols-9 grid-rows-9 gap-1 h-full w-full"
        >
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
          <div
            class="bg-gray-200"
          />
        </div>
      </div>
    </div>
    <div>
      <div
        class="flex flex-col gap-4 justify-center items-center pb-6 text-gray-600"
      >
        <div
          class="text-center"
        >
          Use arrow keys to navigate. Press 1-9 to fill in a cell. Press Backspace to clear a cell.
        </div>
        <div
          class="flex flex-row gap-4 items-center justify-even"
        >
          <button
            data-testid="reset"
          >
            Reset (R)
          </button>
          <button
            data-testid="new"
          >
            New Game (N)
          </button>
          <button
            data-testid="undo"
          >
            Undo (Z)
          </button>
          <button
            data-testid="toggle"
          >
            Highlight errors (E)
          </button>
        </div>
      </div>
    </div>
  </main>
</div>
`;

exports[`Game renders correctly when board is null 1`] = `
<div>
  <main
    class="flex flex-col h-full bg-indigo-50 p-4"
  >
    <div
      class="p-2"
    >
      <h1
        class="text-5xl text-center font-semibold text-indigo-800"
      >
        SUDOKU
      </h1>
      <div
        class="text-center text-gray-600 mt-3"
      >
        Fill the 9x9 grid with digits so that each column, row, and 3x3 section contains all the digits from 1 to 9.
      </div>
    </div>
    <div
      class="grow flex flex-col justify-center items-center mt-4 mb-8"
    >
      <div
        class="aspect-square h-full"
        style="max-height: 75vh; max-width: 80vw;"
      >
        <div
          class="grid grid-cols-9 grid-rows-9 h-full w-full border-4 border-violet-200 rounded-xl overflow-hidden"
        >
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2 border-b-2"
            data-testid="cell"
            id="cell-0"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="8"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2 border-b-2"
            data-testid="cell"
            id="cell-1"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="3"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-4 border-b-2"
            data-testid="cell"
            id="cell-2"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="7"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2 border-b-2"
            data-testid="cell"
            id="cell-3"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="6"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2 border-b-2"
            data-testid="cell"
            id="cell-4"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="2"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-4 border-b-2"
            data-testid="cell"
            id="cell-5"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="9"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2 border-b-2"
            data-testid="cell"
            id="cell-6"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="1"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2 border-b-2"
            data-testid="cell"
            id="cell-7"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="4"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-b-2"
            data-testid="cell"
            id="cell-8"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="5"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium text-indigo-400 bg-gray-100 border-violet-200 border-r-2 border-b-2"
            data-testid="cell"
            id="cell-9"
            max="9"
            min="1"
            step="1"
            type="number"
            value=""
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2 border-b-2"
            data-testid="cell"
            id="cell-10"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="4"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium text-indigo-400 bg-gray-100 border-violet-200 border-r-4 border-b-2"
            data-testid="cell"
            id="cell-11"
            max="9"
            min="1"
            step="1"
            type="number"
            value=""
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2 border-b-2"
            data-testid="cell"
            id="cell-12"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="3"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2 border-b-2"
            data-testid="cell"
            id="cell-13"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="1"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-4 border-b-2"
            data-testid="cell"
            id="cell-14"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="8"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium text-indigo-400 bg-gray-100 border-violet-200 border-r-2 border-b-2"
            data-testid="cell"
            id="cell-15"
            max="9"
            min="1"
            step="1"
            type="number"
            value=""
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium text-indigo-400 bg-gray-100 border-violet-200 border-r-2 border-b-2"
            data-testid="cell"
            id="cell-16"
            max="9"
            min="1"
            step="1"
            type="number"
            value=""
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-b-2"
            data-testid="cell"
            id="cell-17"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="2"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2 border-b-4"
            data-testid="cell"
            id="cell-18"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="9"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2 border-b-4"
            data-testid="cell"
            id="cell-19"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="2"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-4 border-b-4"
            data-testid="cell"
            id="cell-20"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="1"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2 border-b-4"
            data-testid="cell"
            id="cell-21"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="5"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2 border-b-4"
            data-testid="cell"
            id="cell-22"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="7"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-4 border-b-4"
            data-testid="cell"
            id="cell-23"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="4"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2 border-b-4"
            data-testid="cell"
            id="cell-24"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="3"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2 border-b-4"
            data-testid="cell"
            id="cell-25"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="6"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-b-4"
            data-testid="cell"
            id="cell-26"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="8"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium text-indigo-400 bg-gray-100 border-violet-200 border-r-2 border-b-2"
            data-testid="cell"
            id="cell-27"
            max="9"
            min="1"
            step="1"
            type="number"
            value=""
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2 border-b-2"
            data-testid="cell"
            id="cell-28"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="5"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-4 border-b-2"
            data-testid="cell"
            id="cell-29"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="4"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2 border-b-2"
            data-testid="cell"
            id="cell-30"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="1"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2 border-b-2"
            data-testid="cell"
            id="cell-31"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="8"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-4 border-b-2"
            data-testid="cell"
            id="cell-32"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="6"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2 border-b-2"
            data-testid="cell"
            id="cell-33"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="2"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2 border-b-2"
            data-testid="cell"
            id="cell-34"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="3"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-b-2"
            data-testid="cell"
            id="cell-35"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="9"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2 border-b-2"
            data-testid="cell"
            id="cell-36"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="1"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2 border-b-2"
            data-testid="cell"
            id="cell-37"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="6"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-4 border-b-2"
            data-testid="cell"
            id="cell-38"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="3"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium text-indigo-400 bg-gray-100 border-violet-200 border-r-2 border-b-2"
            data-testid="cell"
            id="cell-39"
            max="9"
            min="1"
            step="1"
            type="number"
            value=""
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium text-indigo-400 bg-gray-100 border-violet-200 border-r-2 border-b-2"
            data-testid="cell"
            id="cell-40"
            max="9"
            min="1"
            step="1"
            type="number"
            value=""
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium text-indigo-400 bg-gray-100 border-violet-200 border-r-4 border-b-2"
            data-testid="cell"
            id="cell-41"
            max="9"
            min="1"
            step="1"
            type="number"
            value=""
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2 border-b-2"
            data-testid="cell"
            id="cell-42"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="8"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium text-indigo-400 bg-gray-100 border-violet-200 border-r-2 border-b-2"
            data-testid="cell"
            id="cell-43"
            max="9"
            min="1"
            step="1"
            type="number"
            value=""
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-b-2"
            data-testid="cell"
            id="cell-44"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="7"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2 border-b-4"
            data-testid="cell"
            id="cell-45"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="2"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2 border-b-4"
            data-testid="cell"
            id="cell-46"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="8"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-4 border-b-4"
            data-testid="cell"
            id="cell-47"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="9"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium text-indigo-400 bg-gray-100 border-violet-200 border-r-2 border-b-4"
            data-testid="cell"
            id="cell-48"
            max="9"
            min="1"
            step="1"
            type="number"
            value=""
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2 border-b-4"
            data-testid="cell"
            id="cell-49"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="5"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-4 border-b-4"
            data-testid="cell"
            id="cell-50"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="3"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2 border-b-4"
            data-testid="cell"
            id="cell-51"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="4"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2 border-b-4"
            data-testid="cell"
            id="cell-52"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="1"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-b-4"
            data-testid="cell"
            id="cell-53"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="6"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium text-indigo-400 bg-gray-100 border-violet-200 border-r-2 border-b-2"
            data-testid="cell"
            id="cell-54"
            max="9"
            min="1"
            step="1"
            type="number"
            value=""
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium text-indigo-400 bg-gray-100 border-violet-200 border-r-2 border-b-2"
            data-testid="cell"
            id="cell-55"
            max="9"
            min="1"
            step="1"
            type="number"
            value=""
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-4 border-b-2"
            data-testid="cell"
            id="cell-56"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="2"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2 border-b-2"
            data-testid="cell"
            id="cell-57"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="8"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium text-indigo-400 bg-gray-100 border-violet-200 border-r-2 border-b-2"
            data-testid="cell"
            id="cell-58"
            max="9"
            min="1"
            step="1"
            type="number"
            value=""
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-4 border-b-2"
            data-testid="cell"
            id="cell-59"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="5"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2 border-b-2"
            data-testid="cell"
            id="cell-60"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="6"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium text-indigo-400 bg-gray-100 border-violet-200 border-r-2 border-b-2"
            data-testid="cell"
            id="cell-61"
            max="9"
            min="1"
            step="1"
            type="number"
            value=""
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-b-2"
            data-testid="cell"
            id="cell-62"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="1"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium text-indigo-400 bg-gray-100 border-violet-200 border-r-2 border-b-2"
            data-testid="cell"
            id="cell-63"
            max="9"
            min="1"
            step="1"
            type="number"
            value=""
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium text-indigo-400 bg-gray-100 border-violet-200 border-r-2 border-b-2"
            data-testid="cell"
            id="cell-64"
            max="9"
            min="1"
            step="1"
            type="number"
            value=""
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium text-indigo-400 bg-gray-100 border-violet-200 border-r-4 border-b-2"
            data-testid="cell"
            id="cell-65"
            max="9"
            min="1"
            step="1"
            type="number"
            value=""
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2 border-b-2"
            data-testid="cell"
            id="cell-66"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="2"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2 border-b-2"
            data-testid="cell"
            id="cell-67"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="4"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-4 border-b-2"
            data-testid="cell"
            id="cell-68"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="1"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium text-indigo-400 bg-gray-100 border-violet-200 border-r-2 border-b-2"
            data-testid="cell"
            id="cell-69"
            max="9"
            min="1"
            step="1"
            type="number"
            value=""
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium text-indigo-400 bg-gray-100 border-violet-200 border-r-2 border-b-2"
            data-testid="cell"
            id="cell-70"
            max="9"
            min="1"
            step="1"
            type="number"
            value=""
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-b-2"
            data-testid="cell"
            id="cell-71"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="3"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2"
            data-testid="cell"
            id="cell-72"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="3"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2"
            data-testid="cell"
            id="cell-73"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="1"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-4"
            data-testid="cell"
            id="cell-74"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="8"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2"
            data-testid="cell"
            id="cell-75"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="9"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2"
            data-testid="cell"
            id="cell-76"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="6"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-4"
            data-testid="cell"
            id="cell-77"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="7"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2"
            data-testid="cell"
            id="cell-78"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="5"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200 border-r-2"
            data-testid="cell"
            id="cell-79"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="2"
          />
          <input
            class="flex items-center justify-center text-center outline-none font-medium select-none text-indigo-800 bg-gray-100 border-violet-200"
            data-testid="cell"
            id="cell-80"
            max="9"
            min="1"
            readonly=""
            step="1"
            type="number"
            value="4"
          />
        </div>
      </div>
    </div>
    <div>
      <div
        class="flex flex-col gap-4 justify-center items-center pb-6 text-gray-600"
      >
        <div
          class="text-center"
        >
          Use arrow keys to navigate. Press 1-9 to fill in a cell. Press Backspace to clear a cell.
        </div>
        <div
          class="flex flex-row gap-4 items-center justify-even"
        >
          <button
            data-testid="reset"
          >
            Reset (R)
          </button>
          <button
            data-testid="new"
          >
            New Game (N)
          </button>
          <button
            data-testid="undo"
          >
            Undo (Z)
          </button>
          <button
            data-testid="toggle"
          >
            Highlight errors (E)
          </button>
        </div>
      </div>
    </div>
  </main>
</div>
`;
